### PR TITLE
Added an implementation of the JSON serializer

### PR DESCRIPTION
**Pull request by alex**

### CODE REVIEW

File: samples/json.bagel [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                )
+            default:
+                # Is this the right API?
+                panic("Unknown type {}".format(field.type.name))
+
+class _JSONTextEncoder:
+    writer: io.Writer
+
+    def encode(self, t: Text):
+        # TODO: not technically correct.
```
Comment (dreid) [inline samples/json.bagel:76]: not best kind of correct.

File: samples/json.bagel [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        self.writer.write(t.encode("utf-8"))
+
+
+class _JSONIntegerEncoder:
+    writer: io.Writer
+
+    def encode(self, t: Int):
+        self.writer.write(t.to_text().encode("utf-8"))
```
Comment (pquerna) [inline samples/json.bagel:88]: one thing to think about, which isn't really the point of your exercise here, but the output of a json encoder should be a bytes type.

Comment (alex) [inline samples/json.bagel:88]: Yup, the `io.Writer` interface it takes has `def write(Bytes)`.

On Fri, Apr 4, 2014 at 2:45 PM, Paul Querna notifications@github.comwrote:

> In samples/json.bagel:
> 
> > -                # Is this the right API?
> > -                panic("Unknown type {}".format(field.type.name
> >   ))
> >   +
> >   +class _JSONTextEncoder:
> > -    writer: io.Writer
> >   +
> > -    def encode(self, t: Text):
> > -        # TODO: not technically correct.
> > -        self.writer.write(t.encode("utf-8"))
> >   +
> >   +
> >   +class _JSONIntegerEncoder:
> > -    writer: io.Writer
> >   +
> > -    def encode(self, t: Int):
> > -        self.writer.write(t.to_text().encode("utf-8"))
> 
> one thing to think about, which isn't really the point of your exercise
> here, but the output of a json encoder should be a bytes type.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/10/files#r11315318
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: samples/json.bagel [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    def __new__(writer: io.Writer):
+        return new(JSONObjectEncoder, _writer=writer)
+
+    def encode(self, obj: T):
+        self._write_start_object()
+        # How is it communicated that the return value is a compile time
+        # constant, and iterating over it is unrolled and the values are also
+        # compile time constants?
+        for field in reflect.get_fields{T}():
```
Comment (pquerna) [inline samples/json.bagel]: ~~ yeah, so people talk about this in "theory" that their reflection is going to be understood by their compiler, and it'll do magic like that, but no one actually does it. ~~ Okay so pypy does it. still. hrm.

Comment (alex) [inline samples/json.bagel]: I think implementing this is easy, it's making it easy for people to
understand and read and write which is hard. It's totally easy to do this
type of metaprogramming in RPython for example.

On Fri, Apr 4, 2014 at 2:51 PM, Paul Querna notifications@github.comwrote:

> In samples/json.bagel:
> 
> > +from std import io
> > +from std.meta import reflect
> > +
> > +
> > +class{T} JSONObjectEncoder:
> > -    _writer: io.Writer
> >   +
> > -    def **new**(writer: io.Writer):
> > -        return new(JSONObjectEncoder, _writer=writer)
> >   +
> > -    def encode(self, obj: T):
> > -        self._write_start_object()
> > -        # How is it communicated that the return value is a compile time
> > -        # constant, and iterating over it is unrolled and the values are also
> > -        # compile time constants?
> > -        for field in reflect.get_fields{T}():
> 
> yeah, so people talk about this in "theory" that their reflection is going
> to be understood by their compiler, and it'll do magic like that -- but no
> one actually does it.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/alex/bagel/pull/10/files#r11315523
> .

## 

"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Pull request merged**